src/dataflow/server.rs

```diff
@@ -229,7 +229,11 @@ where
     })
 }
 
-/// Todo: write what this is.
+/// Map of source ID to per-partition timestamp history.
+///
+/// Timestamp history is a vector of tuples (partition_count, timestamp, offset),
+/// where the correct timestamp for a given offset `x` is the highest timestamp value for
+/// the first offset >= `x`.
```

Comment:
I'm not sure what you mean by "is the highest timestamp value for the first offset >= x?

Reply:
if you have a message with offset "3" and the (timestamp, offset) history is (0, 5), (1, 5), (2, 5), (3, 6) then the correct timestamp is "2" (first offset >= 3 is 5, highest timestamp for offset=5 is 3).

Is this actually correct? If so, I can add that example to the comment.

Reply:
Ah ok, now I see what you mean. Yes that's correct.